libraries/Microsoft.Bot.Builder.LanguageGeneration/TemplatesParser.cs

```diff
@@ -148,19 +148,22 @@ public static IParseTree AntlrParseTemplates(LGResource resource)
         /// <param name="importResolver">Resolver to resolve LG import id to template text.</param>
         /// <param name="expressionParser">Expression parser for parsing expressions.</param>
         /// <param name="cachedTemplates">Give the file path and templates to avoid parsing and to improve performance.</param>
+        /// <param name="parentTemplates">Parent visited Templates.</param>
         /// <returns>new <see cref="Templates"/> entity.</returns>
         public static Templates ParseResource(
             LGResource resource,
             ImportResolverDelegate importResolver = null,
             ExpressionParser expressionParser = null,
-            Dictionary<string, Templates> cachedTemplates = null)
+            Dictionary<string, Templates> cachedTemplates = null,
+            Stack<Templates> parentTemplates = null)
```

Comment:
Let's keep the public interface unchange, and create an internal method with an extra state: visitedTemplates.